# Source inventory
Examine raw files

## 1. Load files

In [1]:
import pandas as pd

DATA_DIR = "../data"

attributes = pd.read_csv(f"{DATA_DIR}/features_attributes.csv")
financials = pd.read_csv(f"{DATA_DIR}/features_financials.csv")
clickstream = pd.read_csv(f"{DATA_DIR}/feature_clickstream.csv")
lms = pd.read_csv(f"{DATA_DIR}/lms_loan_daily.csv")

for name, df in [("attributes", attributes), ("financials", financials),
                 ("clickstream", clickstream), ("lms", lms)]:
    print(name, df.shape)

attributes (12500, 6)
financials (12500, 22)
clickstream (215376, 22)
lms (137500, 11)


## 2. Examine rows and column types

In [2]:
attributes.head()

,Customer_ID,Name,Age,SSN,Occupation,snapshot_date
0,CUS_0x1000,Alistair Barrf,18,913-74-1218,Lawyer,2023-05-01
1,CUS_0x1009,Arunah,26,063-67-6938,Mechanic,2025-01-01
2,CUS_0x100b,Shirboni,19,#F%$D@*&8,Media_Manager,2024-03-01
3,CUS_0x1011,Schneyerh,44,793-05-8223,Doctor,2023-11-01
4,CUS_0x1013,Cameront,44,930-49-9615,Mechanic,2023-12-01


In [3]:
attributes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12500 entries, 0 to 12499
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Customer_ID    12500 non-null  object
 1   Name           12500 non-null  object
 2   Age            12500 non-null  object
 3   SSN            12500 non-null  object
 4   Occupation     12500 non-null  object
 5   snapshot_date  12500 non-null  object
dtypes: object(6)
memory usage: 586.1+ KB


In [4]:
attributes.groupby("SSN")["Customer_ID"].nunique().sort_values(ascending=False).head(10)

SSN
#F%$D@*&8      703
671-45-7622      1
670-93-4185      1
671-00-6399      1
671-03-2737      1
671-03-8487      1
671-04-1709      1
671-14-3613      1
671-32-4609      1
671-34-5569      1
Name: Customer_ID, dtype: int64

In [5]:
attributes.groupby("Name")["Customer_ID"].nunique().sort_values(ascending=False).head(10)

Name
Stevex                6
Jessicad              6
Langep                6
Andreaj               5
Deepa Seetharamanm    5
Nate Raymondw         5
Ronald Groverk        5
Valetkevitchv         5
Davidc                5
Kimj                  5
Name: Customer_ID, dtype: int64

In [6]:
financials.head()

,Customer_ID,Annual_Income,Monthly_Inhand_Salary,Num_Bank_Accounts,Num_Credit_Card,Interest_Rate,Num_of_Loan,Type_of_Loan,Delay_from_due_date,Num_of_Delayed_Payment,...,Credit_Mix,Outstanding_Debt,Credit_Utilization_Ratio,Credit_History_Age,Payment_of_Min_Amount,Total_EMI_per_month,Amount_invested_monthly,Payment_Behaviour,Monthly_Balance,snapshot_date
0,CUS_0x1000,30625.94,2706.161667,6,5,27,2,"Credit-Builder Loan, and Home Equity Loan",57,26,...,Bad,1562.91,30.077191,10 Years and 9 Months,Yes,42.941090,77.31427572208112,High_spent_Medium_value_payments,400.36080052211616,2023-05-01
1,CUS_0x1009,52312.68_,4250.390000,6,5,17,4,"Not Specified, Home Equity Loan, Credit-Builde...",5,18,...,_,202.68,40.286997,31 Years and 0 Months,Yes,108.366467,58.66019164829086,High_spent_Medium_value_payments,508.01234122645366,2025-01-01
2,CUS_0x100b,113781.38999999998,9549.782500,1,4,1,0,NaN,14,8,...,Good,1030.2,28.592943,15 Years and 10 Months,No,0.000000,617.0792665202719,High_spent_Small_value_payments,597.8989834797281,2024-03-01
3,CUS_0x1011,58918.47,5208.872500,3,3,17,3,"Student Loan, Credit-Builder Loan, and Debt Co...",27,13,...,Standard,473.14,27.829959,15 Years and 10 Months,Yes,123.434939,383.35084463651407,Low_spent_Medium_value_payments,294.1014665671429,2023-11-01
4,CUS_0x1013,98620.98,7962.415000,3,3,6,3,"Student Loan, Debt Consolidation Loan, and Per...",12,9,...,Good,1233.51,26.524864,17 Years and 10 Months,No,228.018084,332.3337079767732,High_spent_Medium_value_payments,485.8897083704929,2023-12-01


In [7]:
financials.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12500 entries, 0 to 12499
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Customer_ID               12500 non-null  object 
 1   Annual_Income             12500 non-null  object 
 2   Monthly_Inhand_Salary     12500 non-null  float64
 3   Num_Bank_Accounts         12500 non-null  int64  
 4   Num_Credit_Card           12500 non-null  int64  
 5   Interest_Rate             12500 non-null  int64  
 6   Num_of_Loan               12500 non-null  object 
 7   Type_of_Loan              11074 non-null  object 
 8   Delay_from_due_date       12500 non-null  int64  
 9   Num_of_Delayed_Payment    12500 non-null  object 
 10  Changed_Credit_Limit      12500 non-null  object 
 11  Num_Credit_Inquiries      12500 non-null  float64
 12  Credit_Mix                12500 non-null  object 
 13  Outstanding_Debt          12500 non-null  object 
 14  Credit

In [8]:
clickstream.head()

,fe_1,fe_2,fe_3,fe_4,fe_5,fe_6,fe_7,fe_8,fe_9,fe_10,...,fe_13,fe_14,fe_15,fe_16,fe_17,fe_18,fe_19,fe_20,Customer_ID,snapshot_date
0,63,118,80,121,55,193,111,112,-101,83,...,-16,-81,-126,114,35,85,-73,76,CUS_0x1037,2023-01-01
1,-108,182,123,4,-56,27,25,-6,284,222,...,-14,-96,200,35,130,94,111,75,CUS_0x1069,2023-01-01
2,-13,8,87,166,214,-98,215,152,129,139,...,26,86,171,125,-130,354,17,302,CUS_0x114a,2023-01-01
3,-85,45,200,89,128,54,76,51,61,139,...,172,96,174,163,37,207,180,118,CUS_0x1184,2023-01-01
4,55,120,226,-86,253,97,107,68,103,126,...,76,43,183,159,-26,104,118,184,CUS_0x1297,2023-01-01


In [9]:
lms.head()

,loan_id,Customer_ID,loan_start_date,tenure,installment_num,loan_amt,due_amt,paid_amt,overdue_amt,balance,snapshot_date
0,CUS_0x1000_2023_05_01,CUS_0x1000,2023-05-01,10,0,10000,0.0,0.0,0.0,10000.0,2023-05-01
1,CUS_0x1000_2023_05_01,CUS_0x1000,2023-05-01,10,1,10000,1000.0,1000.0,0.0,9000.0,2023-06-01
2,CUS_0x1000_2023_05_01,CUS_0x1000,2023-05-01,10,2,10000,1000.0,1000.0,0.0,8000.0,2023-07-01
3,CUS_0x1000_2023_05_01,CUS_0x1000,2023-05-01,10,3,10000,1000.0,0.0,1000.0,8000.0,2023-08-01
4,CUS_0x1000_2023_05_01,CUS_0x1000,2023-05-01,10,4,10000,1000.0,2000.0,0.0,6000.0,2023-09-01


In [10]:
lms.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137500 entries, 0 to 137499
Data columns (total 11 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   loan_id          137500 non-null  object 
 1   Customer_ID      137500 non-null  object 
 2   loan_start_date  137500 non-null  object 
 3   tenure           137500 non-null  int64  
 4   installment_num  137500 non-null  int64  
 5   loan_amt         137500 non-null  int64  
 6   due_amt          137500 non-null  float64
 7   paid_amt         137500 non-null  float64
 8   overdue_amt      137500 non-null  float64
 9   balance          137500 non-null  float64
 10  snapshot_date    137500 non-null  object 
dtypes: float64(4), int64(3), object(4)
memory usage: 11.5+ MB


**Findings:**
- LMS is monthly, not daily: one row per loan per month, with `installment_num` counting months since the loan started
- Several numeric-sounding columns are stored as text because some values are not clean numbers (e.g. `52312.68_`). Affected: `Age`, `Annual_Income`, `Num_of_Loan`, `Num_of_Delayed_Payment`, `Changed_Credit_Limit`,`Outstanding_Debt`, `Amount_invested_monthly`, `Monthly_Balance`
- Only `Type_of_Loan` has real nulls (1,426). Other missing values are written as text placeholders (`_` in `Credit_Mix`; one placeholder SSN shared by 703 customers), so `isna()` does not find them
- Names repeat across up to 6 customers and carry stray letters. Identifiers (`Customer_ID`, `Name`, `SSN`) will not be features.


## 3. Candidate keys

In [11]:
def check_key(df, key):
    """
    Check df against a primary key or composite key by checking 
    for total row count, missing values, and duplicates.
    """
    print("rows:", len(df))
    
    # Check the key columns for missing values
    print("blank key rows:", df[key].isna().any(axis=1).sum())
    
    # Count how many rows share the exact same key values as an earlier row
    print("duplicate key rows:", df.duplicated(subset=key).sum())

for name, df in [("attributes", attributes), ("financials", financials), ("clickstream", clickstream)]:
    print(f"--- {name}")
    check_key(df, ["Customer_ID", "snapshot_date"])
    print("distinct customers:", df["Customer_ID"].nunique())

--- attributes
rows: 12500
blank key rows: 0
duplicate key rows: 0
distinct customers: 12500
--- financials
rows: 12500
blank key rows: 0
duplicate key rows: 0
distinct customers: 12500
--- clickstream
rows: 215376
blank key rows: 0
duplicate key rows: 0
distinct customers: 8974


**Findings:** 
- No blank or duplicate keys
- Attributes and financials hold one row per customer (12,500 each), not monthly snapshots
- Clickstream holds 24 monthly rows for each of 8,974 customers
- LMS holds 12,500 loans, one per customer

## 4. Dates

In [12]:
def check_dates(values):
    """
    Check date strings range, uniqueness and format
    """
    
    # Attempt to convert the series to datetime objects using the specific format. Convert to NaT if error.
    parsed = pd.to_datetime(values, format="%Y-%m-%d", errors="coerce")
    
    # Print statistics
    print("min:", values.min(), " max:", values.max(), " distinct:", values.nunique())
    
    # Count missing/invalid data:
    # 1. Missing values (NaN, None)
    # 2. Not a valid YYYY-MM-DD date
    print("blank:", values.isna().sum(), " not a valid YYYY-MM-DD date:", (parsed.isna() & values.notna()).sum())

for name, df in [("attributes", attributes), ("financials", financials), ("clickstream", clickstream), ("lms", lms)]:
    print(f"--- {name} snapshot_date")
    check_dates(df["snapshot_date"])

print("--- lms loan_start_date")
check_dates(lms["loan_start_date"])

--- attributes snapshot_date
min: 2023-01-01  max: 2025-01-01  distinct: 25
blank: 0  not a valid YYYY-MM-DD date: 0
--- financials snapshot_date
min: 2023-01-01  max: 2025-01-01  distinct: 25
blank: 0  not a valid YYYY-MM-DD date: 0
--- clickstream snapshot_date
min: 2023-01-01  max: 2024-12-01  distinct: 24
blank: 0  not a valid YYYY-MM-DD date: 0
--- lms snapshot_date
min: 2023-01-01  max: 2025-11-01  distinct: 35
blank: 0  not a valid YYYY-MM-DD date: 0
--- lms loan_start_date
min: 2023-01-01  max: 2025-01-01  distinct: 25
blank: 0  not a valid YYYY-MM-DD date: 0


In [13]:
attributes["snapshot_date"].value_counts().sort_index()

snapshot_date
2023-01-01    530
2023-02-01    501
2023-03-01    506
2023-04-01    510
2023-05-01    521
2023-06-01    517
2023-07-01    471
2023-08-01    481
2023-09-01    454
2023-10-01    487
2023-11-01    491
2023-12-01    489
2024-01-01    485
2024-02-01    518
2024-03-01    511
2024-04-01    513
2024-05-01    491
2024-06-01    498
2024-07-01    505
2024-08-01    543
2024-09-01    493
2024-10-01    456
2024-11-01    488
2024-12-01    515
2025-01-01    526
Name: count, dtype: int64

In [14]:
clickstream["snapshot_date"].value_counts().sort_index()

snapshot_date
2023-01-01    8974
2023-02-01    8974
2023-03-01    8974
2023-04-01    8974
2023-05-01    8974
2023-06-01    8974
2023-07-01    8974
2023-08-01    8974
2023-09-01    8974
2023-10-01    8974
2023-11-01    8974
2023-12-01    8974
2024-01-01    8974
2024-02-01    8974
2024-03-01    8974
2024-04-01    8974
2024-05-01    8974
2024-06-01    8974
2024-07-01    8974
2024-08-01    8974
2024-09-01    8974
2024-10-01    8974
2024-11-01    8974
2024-12-01    8974
Name: count, dtype: int64

In [15]:
lms["snapshot_date"].value_counts().sort_index()

snapshot_date
2023-01-01     530
2023-02-01    1031
2023-03-01    1537
2023-04-01    2047
2023-05-01    2568
2023-06-01    3085
2023-07-01    3556
2023-08-01    4037
2023-09-01    4491
2023-10-01    4978
2023-11-01    5469
2023-12-01    5428
2024-01-01    5412
2024-02-01    5424
2024-03-01    5425
2024-04-01    5417
2024-05-01    5391
2024-06-01    5418
2024-07-01    5442
2024-08-01    5531
2024-09-01    5537
2024-10-01    5502
2024-11-01    5501
2024-12-01    5531
2025-01-01    5539
2025-02-01    5028
2025-03-01    4515
2025-04-01    4024
2025-05-01    3526
2025-06-01    3021
2025-07-01    2478
2025-08-01    1985
2025-09-01    1529
2025-10-01    1041
2025-11-01     526
Name: count, dtype: int64

**Findings:** 
- All dates are valid month-starts, so every source is monthly
- Attributes/financials run 2023-01 to 2025-01, clickstream 2023-01 to 2024-12, LMS snapshots 2023-01 to 2025-11, and loan starts 2023-01 to 2025-01
- Attributes' monthly counts match LMS's new loans per month (e.g. 530 in 2023-01), suggesting each customer's attributes/financials are taken in their loan application month (TBC)
- Clickstream covers the same 8,974 customers every month, including months after their loan started

## 5. Rows per customer and per loan

In [16]:
clickstream.groupby("Customer_ID").size().describe()

count    8974.0
mean       24.0
std         0.0
min        24.0
25%        24.0
50%        24.0
75%        24.0
max        24.0
dtype: float64

In [17]:
lms.groupby("loan_id").size().describe()

count    12500.0
mean        11.0
std          0.0
min         11.0
25%         11.0
50%         11.0
75%         11.0
max         11.0
dtype: float64

In [18]:
lms.groupby("Customer_ID")["loan_id"].nunique().max()

np.int64(1)

**Findings:** 
- Every clickstream customer has all 24 months; Up to 3,526 customers have no clickstream at all
- Every loan has 11 monthly rows (start month plus a 10-month tenure), and each customer has exactly one loan
